File: question_3/backend/app/models/recipe.rb
class Recipe < ActiveRecord::Base
  validates :name, presence: true, uniqueness: true
  validates :number_of_people, presence: true
  validates :cooking_time, presence: true
  validates :category, presence: true
  has_many :materials, dependent: :delete_all
  has_many :cooking_processes, dependent: :delete_all
  has_many :nutritions, dependent: :delete_all

  def find_by_name(name)
    find_by(name: name)
  end

  def find_by_id(id)
    find_by(id: id)
  end
end

File: question_3/mobile_app/lib/models/recipe.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:mobile_app/models/cooking_process.dart';
import 'package:mobile_app/models/ingredient.dart';
import 'package:mobile_app/models/nutrition.dart';

part 'recipe.freezed.dart';
part 'recipe.g.dart';

@freezed
abstract class Recipe with _$Recipe {
  const factory Recipe({
    required String id,
    required String name,
    @JsonKey(name: 'how_to_make') String? howToMake,
    @JsonKey(name: 'number_of_people') required int numberOfPeople,
    @JsonKey(name: 'cooking_time') required int cookingTime,
    required String category,
    int? calories,
    List<Ingredient>? materials,
    @JsonKey(name: 'cooking_processes') List<CookingProcess>? cookingProcesses,
    List<Nutrition>? nutritions,
  }) = _Recipe;
  factory Recipe.fromJson(Map<String, dynamic> json) => _$RecipeFromJson(json);
}